### ISSUE TITLE
Invalid PreKey bundles and malformed PreKeyMessages

### ISSUE THREAD

**Issue by chrisballinger** [signalapp/libsignal-protocol-c#65]
I have:
- [X] searched open and closed issues for duplicates

----------------------------------------

### Bug description

After updating to the latest master branch of libsignal-protocol-c I have seen a huge increase in the number of invalid bundles generated, and bad MACs on messages. I cannot definitely point the problem at this library and not some (yet to be determined) external cause, but I have a strong feeling that something is amiss. I will be rolling back to an earlier version to see if this helps, but this problem may have existed in the past (explaining some user reports) and I wasn't able to reproduce it as reliably.

First I thought that I would simply check if the generated bundle was invalid, so I added a utility function that borrowed the sanity checks from `session_builder_process_pre_key_bundle`.

```objc
/** This will do a rough check if bundle is considered valid */
- (BOOL) checkValidity:(NSError **)error {
    // session_builder.c:191
    // int session_builder_process_pre_key_bundle(session_builder *builder, session_pre_key_bundle *bundle)
    
```

Bundles stored for any period of time seem to fail validation, which wasn't the case before updating to latest master. Clients would occasionally publish invalid bundles to the server, preventing sessions from being created, so bundles are a lot fresher lately.

Which brings me to the next problem, that I hadn't seen before. Now I am very frequently seeing "Bad MAC" errors being returned from libsignal-protocol-c when receiving PreKeyMessages. My theory was that they were caused by starting a session with an invalid bundle, but now that invalid bundles are never uploaded to the server, I am not sure what is happening.

One question I just realized that may be related. If the SignedPreKey changes, do all existing PreKeys need to be invalided as well?

The biggest problem is that I can't tell from the outgoing side that the session is corrupted, so once it gets broken it never is able to fix itself, and there is no way to know the receiving side couldn't decrypt other than lack of delivery receipt.

### Steps to reproduce

I don't have a unit test strictly limited to libsignal-protocol-c itself, so I know it is difficult to accept a bug report of nebulous origin, but I really haven't been able to determine a cause for an increase in these errors other than updating the library.

Here is a failing unit test that used to pass before updating the library and adding additional bundle validation checks: https://github.com/ChatSecure/ChatSecure-iOS/blob/78d0898c1d2103ee3127760260c3d9ec80869632/ChatSecureTests/OTROmemoStorageTest.swift#L174

**Actual result:** Generated bundles are sometimes corrupted. Message MACs are sometimes broken.
**Expected result:** Bundles and message MACs always validate.

### Link to logs

```
2017-04-25 15:47:33.681 ChatSecure[95818:6196770] SignalProtocol (2): Bad MAC
2017-04-25 15:47:33.681 ChatSecure[95818:6196770] SignalProtocol (1): Message mac not verified
2017-04-25 15:47:33.686 ChatSecure[95818:6196770] SignalProtocol (2): Bad MAC
2017-04-25 15:47:33.686 ChatSecure[95818:6196770] SignalProtocol (1): Message mac not verified
2017-04-25 15:47:33.687 ChatSecure[95818:6196770] SignalProtocol (1): No valid sessions
```

**Comment by dkonigsberg** [comment]
It would be extremely helpful if you could narrow this down to something that can be reproduced by a unit test against libsignal-protocol-c itself, though that's probably obvious.
If you do manage to show that rolling back the library version makes your problems go away, then I'd like to know the version you rolled back to so I can look at differences.

There really haven't been any changes to the pre_key handling code in quite some time, but its always possible that a change somewhere else has had an unintended side-effect.

**Comment by chrisballinger** [comment]
Thanks for the quick reply! Ideally I will be able to narrow it down further to extract a failing unit test. I wasn't able to reproduce it in the past, so being able to consistently reproduce helps at least. I'll update this issue with my findings after rolling back.

**Comment by dkonigsberg** [comment]
For what its worth, I was actually able to checkout your source code, get the tests to build in Xcode, and see the test failure for myself.  (No small feat, given the size of the project and its dependencies, but I had surprisingly little trouble.)
Unfortunately, I don't have much experience in Swift, Obj-C, or the Xcode debugger. So it seemed like parts of your "bundle" object were changing as they were passed between various load/save functions (which didn't touch this library in any obvious way). But it wasn't obvious how that could happen, or if I was even seeing what I thought I was seeing in the debugger.

**Comment by chrisballinger** [comment]
@dkonigsberg Wow I'm impressed you got that up and running so quickly! Onboarding is a pain 😞. I think you're correct about bundle corruption happening at the storage layer. Since I added code to revalidate bundles when fetching from storage, we still shouldn't be publishing bad bundles, unless perhaps the re-generation step is causing problems.

So far I've tried:
1. Rolling back the library to a previously used version, specifically a1df1a642281c804a95d144b8aa934c0de0ebda1
2. Refactoring to prevent collisions of SignalAddresses between accounts. 

So far no dice. Previously the signal data store was shared among all accounts, but the SignalAddresses could possibly collide if multiple accounts had buddies that shared the same identifier, or you had yourself on your own roster. That refactor didn't seem to help either.

Thanks for checking it out. I'll test out some more hypotheses tomorrow.

**Comment by dkonigsberg** [comment]
I know it won't help you narrow down this specific issue, but one thing you should really have is a way for the receiver to communicate decrypt errors to the sender.  Then you can have the sender retry the encryption in increasingly aggressive ways (from simple retry, to using a different pre-key, to trying to delete and recreate the session state) until you either succeed or give up.  Doing this will allow you to recover from errors like this.

**Comment by chrisballinger** [comment]
@dkonigsberg I think I narrowed down some of the problems. When I added code to manually validate (stored) bundles before re-publishing, there was a bug that indicated that the bundle was corrupted 100% of the time, which would trigger a bundle regeneration. When the bundle was regenerated it would also create a new SignedPreKey but the original design only allowed storage of one SignedPreKey per account. By replacing the single SignedPreKey entry upon bundle regeneration, it would invalidate all prior sessions(?) and cause other problems. Attempting to fix the invalid bundle bug revealed a much worse bug. I needed to do a big refactor to fix this oversight, and am in the process of cleaning that up.

The [latest OMEMO spec](https://xmpp.org/extensions/xep-0384.html) doesn't currently have a way to indicate that the session is corrupted, but IIRC the original draft had a way to refresh a session. I'm pretty sure it's currently broken in a way that will necessitate adding decryption error feedback to fix it.
